src/transform/src/lib.rs

```diff
@@ -437,8 +452,9 @@ pub struct Optimizer {
 
 impl Optimizer {
     /// Builds a logical optimizer that only performs logical transformations.
-    pub fn logical_optimizer() -> Self {
+    pub fn logical_optimizer(ctx: &crate::typecheck::Context) -> Self {
         let transforms: Vec<Box<dyn crate::Transform>> = vec![
+            Box::new(crate::typecheck::Typecheck::new(ctx.clone())),
```

Comment:
I think that'd be a great idea---I started coarse grained just to reduce the noise as I debugged. Finer-grained checking will help us find errors that might be masked (i.e., one optimization breaks types and the next restores it), but is of course much more costly, and we're already at 2%. 

We could address some of this cost by memoizing things.

Semi-related: is there a reason that MRE isn't hashconsed? It would bring the cost of memoization down to zero.

Unrelated: should I have marked this PR as a draft?

Reply:
Hashconsing and CSE are kind of different. CSE (presumably) lifts common sub-expressions into a `let`, which is what will make sure you render code without duplication... if you render hashconsed code without being careful, you'll duplicate any shared nodes.

I had hashconsing in mind here because it's helpful for memoization: you can compute hashes in O(1) (because they're cached, effectively) and so memoization is very cheap.